model/model.go

```diff
@@ -211,32 +211,35 @@ func newCertificate(cert string, key string) *Certificate {
 // SSLConfig represents SSL-related configuration options.
 type SSLConfig struct {
 	Enforce           bool        `key:"enforce" constraint:"(?i)^(true|false)$"`
-	Protocols         string      `key:"protocols" constraint:"^((SSLv2|SSLv3|TLSv1|TLSv1\\.1|TLSv1\\.2)\\s*)+$"`
-	Ciphers           string      `key:"ciphers" constraint:"^(!?[A-Z][A-Z\\d\\+-]+:?)*$"`
+	Protocols         string      `key:"protocols" constraint:"^((SSLv[2-3]|TLSv1(?:\\.[1-3])?)\\s*)+$"`
+	Ciphers           string      `key:"ciphers" constraint:"^((\\b[\\w.!+-]+\\b)+(:?@(STRENGTH|SECLEVEL=[0-5]))?(:([!+-]\\b)?|$))*(((\\b[\\w.+-]+\\b)+|(\\[(\\b[\\w.|+-]+\\b)+\\]))(:|$))*$"`
```

Comment:
Can you give me some examples on how this regex has changed and how the Ciphers need to annotated now?

Reply:
The previous regex was very limited and only matched a small subset of what can be expressed as cipher suites, but existing annotations will continue to work unchanged.

For example it would reject EECDH+AESGCM:EECDH+CHACHA20:+aRSA because it required ciphers to start with an uppercase letter, optionally preceded by an exclamation mark, which isn't even valid as the first term. So eg 3DES would be valid, but rejected, but !DSS would be allowed, but is invalid on its own (but DHE+AES:!DSS would be OK).

You have to take a look at the openssl ciphers manpage/docs to understand all the rules and I tried to encode as many as reasonably possible in a short-ish regex. 

In addition to the openssl ciphers rules it also encodes additional rules that apply when equal preference groups are used, because some more advanced stuff can only be done before the first preference group, as is denoted in the comments of the patch ported from boringssl. 

The patch also allows to list TLS 1.3 ciphers like TLS_AES_128_GCM_SHA256, so the rules had to be extended for that.

In short this regex was a pain in the ass to write and took me hours to get right, especially since go doesn't support lookahead and lookbehind assertions in regular expressions…

Reply:
Well I'm glad you spent time perfecting this regex and everyone will appreciate this new router update soon.